src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EnsureCatalogProductLoadingMiddleware.cs

```diff
@@ -0,0 +1,43 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using PipelineNet.Middleware;
+using VirtoCommerce.CatalogModule.Core.Services;
+using VirtoCommerce.XDigitalCatalog.Queries;
+
+namespace VirtoCommerce.XDigitalCatalog.Middlewares
+{
+    public class EnsureCatalogProductLoadingMiddleware : IAsyncMiddleware<SearchProductResponse>
+    {
+        private readonly IItemService _itemService;
+
+        public EnsureCatalogProductLoadingMiddleware(IItemService itemService)
+        {
+            _itemService = itemService;
+        }
+
+        public async Task Run(SearchProductResponse parameter, Func<SearchProductResponse, Task> next)
+        {
+            var itemsIds = parameter.Results
+                .Where(expProduct => expProduct.IndexedProduct is null)
+                .Select(expProduct => expProduct.Key)
+                .Where(key => key != null)
+                .ToArray();
+
+            if (itemsIds.Any())
+            {
+                var responceGroup = parameter.Query.GetResponseGroup();
```

Comment:
It should be another response group, you need to add the method which calculates `ItemResponseGroup` based on the query fields:
https://github.com/VirtoCommerce/vc-module-catalog/blob/bac3d033464b06eabecc3a9f8a5b1bcc799004f7/src/VirtoCommerce.CatalogModule.Core/Model/ItemResponseGroup.cs#L6